core/store/src/db.rs

```diff
@@ -79,6 +80,119 @@ impl DBTransaction {
 
 pub type DBIterator<'a> = Box<dyn Iterator<Item = io::Result<(Box<[u8]>, Box<[u8]>)>> + 'a>;
 
+/// Data returned from the database.
+///
+/// Abstraction layer for data returned by [`Database::get_raw_bytes`] and
+/// [`Database::get_with_rc_stripped`] methods.  Operating on the value as
+/// a slice is free while converting it to a vector on an arc may requires an
+/// allocation and memory copy.
+pub struct DBBytes<'a>(DBBytesInner<'a>);
+
+enum DBBytesInner<'a> {
+    /// Data held as a vector.
+    Vec(Vec<u8>),
+
+    /// Data held as RocksDB-specific pinnable slice.
+    Rocks {
+        /// Slice view of the data.
+        ///
+        /// Having this allows us to create views at only parts of the pinnable
+        /// slice without having to modify it.  We need it because remove_suffix
+        /// method of the C++ class is not exposed.
+        slice: &'a [u8],
+
+        /// This is what owns the data.
+        ///
+        /// This is never modified which is why we can have `slice` field
+        /// pointing at the data this pinnable slice holds.
+        db_slice: ::rocksdb::DBPinnableSlice<'a>,
+    },
+}
+
+impl<'a> DBBytes<'a> {
+    pub fn as_slice(&self) -> &[u8] {
+        match self.0 {
+            DBBytesInner::Vec(ref bytes) => bytes.as_slice(),
+            DBBytesInner::Rocks { slice, .. } => slice,
+        }
+    }
+
+    fn from_rocksdb_slice(db_slice: ::rocksdb::DBPinnableSlice<'a>) -> Self {
+        let slice = &*db_slice;
+        let (ptr, len) = (slice.as_ptr(), slice.len());
+        // SAFETY: We got the raw parts from a slice so they are obviously
+        // correct.  By using 'a lifetime we guarantee that this slice won’t
+        // outlive `db_slice` field.  And lastly, since `db_slice` is never
+        // modified and it holds pointers to pinned (in Rust sense) data, we can
+        // safely keep reference to its interior.
+        let slice = unsafe { std::slice::from_raw_parts(ptr, len) };
+        DBBytes(DBBytesInner::Rocks { slice, db_slice })
+    }
+
+    fn strip_refcount(self) -> Option<Self> {
+        match self.0 {
+            DBBytesInner::Vec(bytes) => {
+                refcount::strip_refcount(bytes).map(|bytes| Self(DBBytesInner::Vec(bytes)))
+            }
+            DBBytesInner::Rocks { slice, db_slice } => refcount::decode_value_with_rc(slice)
+                .0
+                .map(|slice| Self(DBBytesInner::Rocks { slice, db_slice })),
+        }
+    }
+}
+
+impl<'a> std::ops::Deref for DBBytes<'a> {
+    type Target = [u8];
+    fn deref(&self) -> &[u8] {
+        self.as_slice()
+    }
+}
+
+impl<'a> From<Vec<u8>> for DBBytes<'a> {
+    fn from(bytes: Vec<u8>) -> Self {
+        Self(DBBytesInner::Vec(bytes))
+    }
+}
+
+impl<'a> From<&[u8]> for DBBytes<'a> {
+    fn from(bytes: &[u8]) -> Self {
+        Self(DBBytesInner::Vec(bytes.to_vec()))
+    }
+}
```

Comment:
Not a fan of this impl, as it hides a copy.

Reply:
I was inspired by [analogous Vec implementation](https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-From%3C%26%5BT%5D%3E).

Reply:
Yeah, I can't say I am a huge fan of that impl either. 

There's a tradeoff between making the costs explicit, and being more re-usable in generic context. 

As Vec is vocabularly type, genericness matter relatively more. In contracst, the reason for DBBytes to exist is to avoid needless copies, so cost explicitness looms larger. I don't think we even use this conversion in a generic way?

So I'd say it's right for Vec to have this impl, and for DBBytes to not have on.

The original motivation explained in https://github.com/rust-lang/rfcs/blob/master/text/0529-conversion-traits.md

> Use ad hoc conversion methods for "natural", outgoing conversions that should have easy method names and good discoverability. A conversion is "natural" if you'd call it directly on the type in normal code; "unnatural" conversions usually come from generic programming.
>
> For example, to_string is a natural conversion for str, while into_string is not; but the latter is sometimes useful in a generic context -- and that's what the generic conversion traits can help with.


No strong opinion overall!